### PR TITLE
Implementation of operators for B0 distortion correction 

### PR DESCRIPTION
**Pull request by schote**
Edit by @fzimmermann89 

This adds three new operators
- Time Segmented Fourier 
- Multi Frequency Fourier
- Conjugate Phase Fourier

utility code for random generation of spherical harmonics

example notbook

It also includes a fix for RandomGenerator without seed

Original content:

> This PR aims to implement two new operators for B0 distortion correction.
> Both operators rely on the implemented FFT operator.
> 
> `TimeSegmentedFastFourierOp.py`
> - Uses all frequencies in the provided field map or a maximum of `num_frequencies` based on the distribution of frequencies in the field map
> - Performs a segmentation of the readout times and calculates an interpolator
> 
> `ConjugatePhaseFastFourierOp.py`
> - Considers all unique frequencies in the provided field map
> - Calculates full phase term with all readout time points
> 
> Examplary [notebook](https://github.com/PTB-MR/mrpro/blob/06a6939990eb70468a11a9348c669d6a85698069/examples/notebooks/b0_distortion_correction_example.ipynb) demonstrages usage of both operators.

### CODE REVIEW

**Comment by fzimmermann89** [comment]
Nice! Just skimming over it, a few remarks:

- Fourier Operator: These should receive a Fourier operator in the __init__. You currently initialize self._fft_op = FastFourierOp() with default settings, which only works for fully sampled Cartesian data.

- Einops: Consider using einops.einsum to allow multi-letter axis names. This would make comments like # lzyx Basis (L, Z, Y, X) obsolete.

- Variable Naming: t_ro = readout_times — why not just keep readout_times? Descriptive variable names replace the need for comments.

- Docstrings & Hints: Check the docstrings of the other operators and include a __call__ wrapper for type hinting.

- Conjugate Phase: Do we actually need conjugate phase? It performs one Fourier transform per unique value; is this ever used in practice?

- Testing: If we keep it, can we add a test to ensure results match approximately between the two algorithms?

**Comment by schote** [comment]
> Nice! Just skimming over it, a few remarks:
> 
> * Fourier Operator: These should receive a Fourier operator in the **init**. You currently initialize self._fft_op = FastFourierOp() with default settings, which only works for fully sampled Cartesian data.
> * Einops: Consider using einops.einsum to allow multi-letter axis names. This would make comments like # lzyx Basis (L, Z, Y, X) obsolete.
> * Variable Naming: t_ro = readout_times — why not just keep readout_times? Descriptive variable names replace the need for comments.
> * Docstrings & Hints: Check the docstrings of the other operators and include a **call** wrapper for type hinting.
> * Conjugate Phase: Do we actually need conjugate phase? It performs one Fourier transform per unique value; is this ever used in practice?
> * Testing: If we keep it, can we add a test to ensure results match approximately between the two algorithms?

Thanks for the quick feedback @fzimmermann89 ! 
The implementation of the conjugate phase operator was mostly motivated by having a test for the time segmented operator.

**Comment by schote** [comment]
Example reconstruction using a B0-uninformed FFT operator and the implemented B0-informed operator. The kspace data was simulated using mrzero.
c24a414e-89a3-4d8a-9180-0461262b683d
Ground truth field map (values in Hz):
6816022d-b7d3-46f2-8671-c4306ba5972a

File: src/mrpro/operators/B0InformedFourierOp.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+            raise ValueError('readout_times must have same size as last dim of b0_map')
+
+        device = b0_map.device
+        self._fft_op: FastFourierOp = fft_op.to(device)
+        b0_map_rad = torch.round(2 * torch.pi * b0_map, decimals=b0_decimals)  # Convert to radians/sec
+
+        # Mode 1: Time-Segmented Approximation
+        # Solves for an interpolator (C) such that: Basis(t_seg) @ C ≈ Basis(t_readout)
+        if num_time_points is not None and num_time_points > 0:
+            # Define Time Segments
```
Comment (fzimmermann89) [inline src/mrpro/operators/B0InformedFourierOp.py]: lets get rid of unnessery comment like this one.

File: src/mrpro/operators/B0InformedFourierOp.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        If only `num_frequencies` is set, conjugate phase reconstruction with frequency binning is performed.
+        If both `num_time_points` and `num_frequencies` are set, time-segmented approximation with frequency
+        binning is applied.
+    """
+
+    def __init__(
+        self,
+        b0_map: torch.Tensor,
+        readout_times: torch.Tensor,
+        fft_op: FastFourierOp,
```
Comment (fzimmermann89) [inline src/mrpro/operators/B0InformedFourierOp.py]: this should work with a fourierop

File: src/mrpro/operators/B0InformedFourierOp.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        num_time_points: int | None = None,
```
Comment (fzimmermann89) [inline src/mrpro/operators/B0InformedFourierOp.py]: n_something instead of num_something accoring to our style guide

File: src/mrpro/operators/B0InformedFourierOp.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            Number of decimals to round B0 map values to after converting to radians/sec.
+        """
+        super().__init__()
+
+        # Validate inputs
+        if readout_times.shape[-1] != b0_map.shape[-1]:
+            raise ValueError('readout_times must have same size as last dim of b0_map')
+
+        device = b0_map.device
+        self._fft_op: FastFourierOp = fft_op.to(device)
```
Comment (fzimmermann89) [inline src/mrpro/operators/B0InformedFourierOp.py]: no need to move the operator

File: src/mrpro/operators/B0InformedFourierOp.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+            if num_frequencies is not None and num_frequencies > 0:
+                # Quantile Binning
+                q_steps = torch.linspace(0, 1, num_frequencies + 1, device=device)
+                boundaries = torch.quantile(b0_map_rad.flatten(), q_steps)
+
+                # Use bin centers for frequency evolution
+                frequencies_rad = (boundaries[:-1] + boundaries[1:]) / 2
+
+                # Quantize/bucketize the map, indices will range [0, num_frequencies-1]
+                indices = torch.bucketize(b0_map_rad, boundaries[:-1], right=True) - 1
```
Comment (fzimmermann89) [inline src/mrpro/operators/B0InformedFourierOp.py]: this is not differentiable wrt b0. is this on purpose? can we lienar interpolate instead between the bucket edges or somthing? or we should at least mention it in the docstring and raise a warning if b0.requiees_grad

File: src/mrpro/operators/B0InformedFourierOp.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+            # Compute Basis (One-Hot Masks) with shape: (number of frequencies/bins, *b0map shape)
+            # WARNING: This can be memory intensive if number of bins, i.e. the number of frequencies is large
+            idx_range = torch.arange(num_bins, device=device).view(-1, *([1] * b0_map_rad.ndim))
+            spatial_basis = (indices.unsqueeze(0) == idx_range).to(dtype=torch.complex64)
+
+        # Register buffers
+        self._spatial_basis: torch.Tensor
+        self._temporal_basis: torch.Tensor
+        self.register_buffer('_spatial_basis', spatial_basis)  # Time-Seg -> Complex Maps | Binning -> Boolean Masks
```
Comment (fzimmermann89) [inline src/mrpro/operators/B0InformedFourierOp.py]: no need for register buffer, for oeprators you can just do oeprator.blub=tensor, the setattr is overwritten

File: src/mrpro/operators/B0InformedFourierOp.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+        num_time_points: int | None = None,
+        num_frequencies: int | None = None,
+        b0_decimals: int = 0,
+    ) -> None:
+        """Initialize B0 informed Fourier operator.
+
+        Parameters
+        ----------
+        b0_map
+            Off-resonance map in Hz. Shape (..., Z, Y, X).
```
Comment (fzimmermann89) [inline src/mrpro/operators/B0InformedFourierOp.py]: consistency: use all lowere case letters for z,y,x

File: src/mrpro/operators/B0InformedFourierOp.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+    """
+
+    def __init__(
+        self,
+        b0_map: torch.Tensor,
+        readout_times: torch.Tensor,
+        fft_op: FastFourierOp,
+        num_time_points: int | None = None,
+        num_frequencies: int | None = None,
+        b0_decimals: int = 0,
```
Comment (fzimmermann89) [inline src/mrpro/operators/B0InformedFourierOp.py]: why? why in decimals?

File: src/mrpro/operators/B0InformedFourierOp.py [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+                frequencies_rad, indices = torch.unique(b0_map_rad, return_inverse=True)
+                indices = indices.view(b0_map_rad.shape)
+
+            num_bins = len(frequencies_rad)
+
+            # Compute Phasor (Temporal Evolution) with shape: (N_bins, N_readout)
+            temporal_basis = torch.exp(-1j * torch.einsum('l, x -> lx', frequencies_rad, readout_times))
+
+            # Compute Basis (One-Hot Masks) with shape: (number of frequencies/bins, *b0map shape)
+            # WARNING: This can be memory intensive if number of bins, i.e. the number of frequencies is large
```
Comment (fzimmermann89) [inline src/mrpro/operators/B0InformedFourierOp.py]: why this warning here? whom are we warning? and what to do about it?

**Comment by fzimmermann89** [comment]
I have thgouth a bit more about this, maybe it is cleaner to just implement four operators:
A joint abstract base, B0InformedFourierOp.
and then three variants, ConjugatePhase, MFI and TimeSegmented, that calculate the cofficients differently.

Also, we should tests that the result is differentiable wrt b0 in all three cases. So for MFI we would need to use a weighted interpolation instead of hard bin assignment.

**Comment by fzimmermann89** [comment]
.. not sure if this is really correct, but design wise maybe like this: https://github.com/fzimmermann89/mr2/blob/5851658ab889f32de66088871e444fb300ff9884/src/mr2/operators/B0InformedFourierOp.py

File: examples/notebooks/b0_correction.ipynb [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+   "source": []
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7",
+   "metadata": {
+    "lines_to_next_cell": 0
+   },
+   "source": [
+    "## Simulate a distorted k-space data\n",
```
Comment (schote) [inline examples/notebooks/b0_correction.ipynb]:
```suggestion
    "## Simulate distorted k-space data\n",
```

File: examples/notebooks/b0_correction.ipynb [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+    "We simulate a distorted k-space data by applying the B0-informed Fourier operator to the phantom."
```
Comment (schote) [inline examples/notebooks/b0_correction.ipynb]:
```suggestion
    "We simulate distorted k-space data by applying the B0-informed Fourier operator to the phantom."
```

File: examples/notebooks/b0_correction.ipynb [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "11",
+   "metadata": {
+    "lines_to_next_cell": 0
+   },
+   "source": [
+    "## Correct for B0 inhomogeneity\n",
+    "We can use a faster approximation of the B0-informed Fourier operator, i.e., the Time-Segmented operator\n",
```
Comment (schote) [inline examples/notebooks/b0_correction.ipynb]:
```suggestion
    "We can use a faster approximation of the B0-informed Fourier operator, e.g. the Time-Segmented operator\n",
```

File: examples/notebooks/b0_correction.ipynb [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+    "plt.tight_layout()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "13",
+   "metadata": {},
+   "source": [
+    "## Summary\n",
+    "We have explored how to use the B0-informed Fourier operators to generate a distorted k-space data and\n",
```
Comment (schote) [inline examples/notebooks/b0_correction.ipynb]:
```suggestion
    "We have explored how to use the B0-informed Fourier operators to generate distorted k-space data and\n",
```

File: src/mrpro/operators/B0InformedFourierOp.py [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+        )
+
+        if frequency_centers.numel() == 1:
+            spatial_basis = torch.ones(
+                (1, *b0_map_rad.shape), dtype=b0_map_rad.dtype.to_complex(), device=b0_map_rad.device
+            )
+            return spatial_basis, temporal_basis
+
+        n_centers = frequency_centers.numel()
+        b0_map_flat = b0_map_rad.reshape(-1)
```
Comment (schote) [inline src/mrpro/operators/B0InformedFourierOp.py]: `b0_map_rad_flat`

File: src/mrpro/operators/B0InformedFourierOp.py [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+    def _compute_basis(
+        self, b0_map_rad: torch.Tensor, readout_times: torch.Tensor
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        time_segments = torch.linspace(readout_times[0], readout_times[-1], self.n_segments, device=b0_map_rad.device)
+
+        with torch.no_grad():
+            quantile_steps = torch.linspace(
+                0, 1, self.n_design_frequencies, device=b0_map_rad.device, dtype=b0_map_rad.dtype
+            )
+            design_frequencies = torch.unique(torch.quantile(b0_map_rad.flatten(), quantile_steps))
```
Comment (schote) [inline src/mrpro/operators/B0InformedFourierOp.py:226]: Above, the short form `_freq` ist used, maybe change to `design_freqs_rad`

File: src/mrpro/operators/B0InformedFourierOp.py [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+
+            segment_phases = torch.exp(-1j * design_frequencies[:, None] * time_segments[None, :])
+            target_phases = torch.exp(-1j * design_frequencies[:, None] * readout_times[None, :])
+            temporal_basis = torch.linalg.lstsq(segment_phases, target_phases, rcond=1e-15).solution.to(
```
Comment (schote) [inline src/mrpro/operators/B0InformedFourierOp.py:231]: https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=10517636

eq. (5)

**Comment by fzimmermann89** [comment]
@schote thanks for reviewing this. :)



Kind-a-strange, but as you drafted the math then I completly rewrote the code, I think it was great for YOU to review, as you know what thee oeprators are supposed to do 
(instead of me reviewing my own code or sopmebody who does not know what they should do)

I think it is fine to go, and if nobody vetos, I would merge tonight for the next release


(Small tip for the next time: its easier to review/suggest changes/edit in vscode the example script.py files -- the notebook.ipynb is autogenerated anyways.)

**Review by fzimmermann89 — APPROVED** [review]

**Comment by fzimmermann89** [comment]
I really thought #983 would help here -- but its a slightly different format .....
